### PR TITLE
Add support for proto message field naming conventions (snake_case)

### PR DESCRIPTION
**Pull request by BenVercammen**
The message fields in the generated `.proto` files are not following the protocol buffer naming conventions. This pull request adds the `useProtoFieldNamingConventions` boolean option to enable naming convention compliance. 

It can be set on 2 levels:
 - **For ALL messages and enums**: in the `build.yaml` file configuration so it applies to all generated `.proto` files (see included yaml file)
 - **For a SINGLE  message or enum**: using `@Proto(useProtoFieldNamingConventions: true)` on a class will only do so for the annotated class

For my purposes, the current implementation is sufficient, however I'd like to get some feedback on some open questions:
 1. Where best to document this feature? Explicitly mention it in one of the `README` files? Or add it to one of the example files? Since most README files refer to the example project, I'll just add it over there.
 2. Should the `proto_annotations` dependency in `proto_generator` be changed? I assume the `proto_annotations` version should be upped anyway?
 **Edit:** opened https://github.com/squarealfa/dart_framework/pull/20 for this reason
 3. Should the generated "**proto_services**" also be supporting the naming convention? For now I haven't implemented it, as I currently don't have any need for them myself...
 4. For now, the `useProtoFieldNamingConventions` configuration is implemented as an "OR". If it is set in the `build.yaml` file, it cannot be undone by explicitly setting it to `false` in an `@Proto` annotation. Is that a problem?

One thing I'd like to explain:
The newly introduced `forEnum` in `FieldDescriptor` is required, since these are being used for both "class/message" and "enum" fields, but at the critical point, this information wasn't required. At first I tried using the `isEnumType` field, but that would be true in both cases, while we only need to change the name for the values of an enum, not for an enum type in a class.

### CODE REVIEW

**Comment by ruicraveiro** [comment]
Hi, I think you are right about the conventions, so unless you give me a reason for otherwise, I think I'll go even further and make it the default.

**Comment by BenVercammen** [comment]
> Hi, I think you are right about the conventions, so unless you give me a reason for otherwise, I think I'll go even further and make it the default.

There are a few reasons I can think of, not sure if they are actual show-stoppers though:
 1. It breaks backward compatibility, as the generated `.proto` files will have different field names and the enum value names are completely different, so not sure if that might be a problem for existing projects.
 2. The descriptors in the `protoc` generated `.pbjson.dart` are also different. Not sure how or where these are being used, but I can imagine that this might break backwards compatibility on existing projects.

The backwards compatibility issues might be resolved by upping the minor (or even major?) version of the package? But I'm not sure on how the current versioning scheme actually works. So it seemed safer to me to just keep this a non-default option. Maybe make it default in the next major release?

**Comment by ruicraveiro** [comment]
> The backwards compatibility issues might be resolved by upping the minor (or even major?) version of the package? But I'm not sure on how the current versioning scheme actually works. So it seemed safer to me to just keep this a non-default option. Maybe make it default in the next major release?

I prefer to break early than have the wrong default linger. It is not strictly a Dart breaking change (if both server and client run Dart and are generated from this package, all that developers need to do is at most re-run the builder). However, it can be considered a breaking change, indeed, as you mentioned. Therefore, I intend to bump it to version 3.0.0 to signal this.

Dart pub uses semantic versioning: https://semver.org/spec/v2.0.0-rc.1.html

In a nutshell: All versions from 2.0.0 to < 3.0.0 are promised not to break. Version 3.0.0 can introduce breaking changes from 2.0.0. That's why in all my pubspec.yaml files I add dependencies using the caret syntax:
```yaml
package: ^x.y.z
```

This means that whenever I run ```dart pub upgrade```, unless I add the parameter ```--major-versions```, the tool will upgrade to the latest version of all major x versions, but will not automatically upgrade to a version greater than x.

**Comment by ruicraveiro** [comment]
BTW: Check this out: https://github.com/squarealfa/dart_framework/blob/3bd83992d7812f159bec8438835ed264c5442e85/proto_mapper/proto_generator/pubspec.yaml

I have created a new working branch so I could merge your PR's into a single place and work on them as a single entity. In order to, temporarily override the dependencies between the packages, note that I have added a new section: 

```yaml
  dependency_overrides:
    proto_annotations:
      path: ../proto_annotations        
```

You can save yourself some work by doing the same thing next time you need to create a PR that requires changing several dependent packages, thus creating a single PR. I will remove this immediately before finally publishing the package.

Here's more information: https://dart.dev/tools/pub/dependencies#dependency-overrides

**Comment by ruicraveiro** [comment]
I think the same logic applies to the DateTime precision. We shall break it in version 3.0.0 but leave the option for the user to configure an override. In this case it makes even more sense, because the other part of this project, the map_mapper will convert DateTimes into an Iso8601 string, which supports microsecond precision. What do you think?

**Comment by BenVercammen** [comment]
> I think the same logic applies to the DateTime precision. We shall break it in version 3.0.0 but leave the option for the user to configure an override. In this case it makes even more sense, because the other part of this project, the map_mapper will convert DateTimes into an Iso8601 string, which supports microsecond precision. What do you think?

I'm in favor of increasing the precision by default. Not using map_mapper, but I can see the value in keeping those 2 in line.
